File: Clients/BlogBlazor/Shared/Blazor.Contracts/V1/Responses/PostsResponses/PostResponse.cs
namespace Blazor.Contracts.V1.Responses.PostsResponses;

using System;
using System.Collections.Generic;
using CommentsResponses;
using UsersResponses;

/// <summary>
/// Post response.
/// </summary>
public class PostResponse
{
    /// <summary>
    /// Gets or sets title.
    /// </summary>
    public string Title { get; set; }

    /// <summary>
    /// Gets or sets description.
    /// </summary>
    public string Description { get; set; }

    /// <summary>
    /// Gets or sets content.
    /// </summary>
    public string Content { get; set; }

    /// <summary>
    /// Gets or sets seen.
    /// </summary>
    public int Seen { get; set; }

    /// <summary>
    /// Gets or sets likes.
    /// </summary>
    public int Likes { get; set; }

    /// <summary>
    /// Gets or sets dislikes.
    /// </summary>
    public int Dislikes { get; set; }

    /// <summary>
    /// Gets or sets image url.
    /// </summary>
    public string ImageUrl { get; set; }

    /// <summary>
    /// Gets or sets author id.
    /// </summary>
    public string AuthorId { get; set; }

    /// <summary>
    /// Gets or sets application user.
    /// </summary>
    public virtual ApplicationUserResponse Author { get; set; }

    /// <summary>
    /// Gets or sets created at.
    /// </summary>
    public DateTime CreatedAt { get; set; } = DateTime.Now;

    /// <summary>
    /// Gets or sets the comments.
    /// </summary>
    /// <value>
    /// The comments.
    /// </value>
    public virtual ICollection<CommentResponse> Comments { get; set; }

    /// <summary>
    /// Gets or sets the posts tags relations.
    /// </summary>
    /// <value>
    /// The posts tags relations.
    /// </value>
    public virtual ICollection<PostTagRelationsResponse> PostsTagsRelations { get; set; }
}

File: Shared/Blog.FSharp.Contracts/V1/Responses/PostsResponses/PostResponse.fs
namespace Blog.FSharp.Contracts.V1.Responses.PostsResponses

open System
open System.Collections.Generic
open Blog.FSharp.Contracts.V1.Responses.UsersResponses
open Blog.FSharp.Contracts.V1.Responses.CommentsResponses
open Blog.FSharp.Contracts.V1.Responses.TagsResponses

/// <summary>
/// Post response.
/// </summary>
type PostResponse() =
    /// <summary>Gets or sets title.</summary>
    member val Title: string = "" with get, set

    /// <summary>Gets or sets description.</summary>
    member val Description: string = "" with get, set

    /// <summary>Gets or sets content.</summary>
    member val Content: string = "" with get, set

    /// <summary>Gets or sets seen.</summary>
    member val Seen: int = 0 with get, set

    /// <summary>Gets or sets likes.</summary>
    member val Likes: int = 0 with get, set

    /// <summary>Gets or sets dislikes.</summary>
    member val Dislikes: int = 0 with get, set

    /// <summary>Gets or sets image url.</summary>
    member val ImageUrl: string = "" with get, set

    /// <summary>Gets or sets author id.</summary>
    member val AuthorId: string = "" with get, set

    /// <summary>Gets or sets application user.</summary>
    member val Author: ApplicationUserResponse|null = null with get, set

    /// <summary>Gets or sets created at.</summary>
    member val CreatedAt: DateTime = DateTime.Now with get, set

    /// <summary>Gets or sets the comments.</summary>
    member val Comments: ICollection<CommentResponse> = List<CommentResponse>() :> ICollection<CommentResponse> with get, set

    /// <summary>Gets or sets the posts tags relations.</summary>
    member val PostsTagsRelations: ICollection<PostTagRelationsResponse> = List<PostTagRelationsResponse>() :> ICollection<PostTagRelationsResponse> with get, set

/// <summary>
/// Post tag relations response.
/// </summary>
and PostTagRelationsResponse() =
    /// <summary>Gets or sets the post identifier.</summary>
    member val PostId: int = 0 with get, set

    /// <summary>Gets or sets the post.</summary>
    member val Post: PostResponse|null = null with get, set

    /// <summary>Gets or sets the tag identifier.</summary>
    member val TagId: int = 0 with get, set

    /// <summary>Gets or sets the tag.</summary>
    member val Tag: TagResponse|null = null with get, set

